### PR TITLE
Support Keras-style None dynamic dimensions in Input

### PR DESCRIPTION
**Pull request by El3ssar**
This adds Keras-style dynamic dimensions to `Input`. In Keras you write `Input(shape=(None, 128))` to say "this dimension varies"; here you currently have to pass a concrete number, and the model then quietly accepts other sizes at runtime — so the reported shapes are misleading about what's actually fixed.

With this, you can mark a dimension as dynamic with `None`:

```python
inputs = Input(shape=(3, None, None))   # dynamic height and width
inputs.shape                            # (1, 3, None, None)
inputs.H, inputs.W                      # None, None

x = nn.Conv2d(inputs.C, 16, 3, padding=1)(inputs)
model = SymbolicModel(inputs, x)
model(torch.rand(2, 3, 32, 48))         # any H, W works
```

Details:

- Dynamic dims are reported as `None` in `.shape`, the shortcut properties (`.C/.H/.W/.features/.batch_size`) and `model.summary()` — same convention already used for unknown batch size.
- Tracing substitutes a placeholder size (default 16, configurable with the new `dynamic_size_hint` argument for deep downsampling stacks).
- `batch_shape=(None, ...)` marks the batch size as unknown explicitly.
- Inputs without any `None` are completely unaffected — `.shape` still returns the same `torch.Size`.
- Reporting `None` is intentionally the fail-loud choice: `nn.Linear(x.shape[1], ...)` on a dynamic dim now raises at trace time instead of silently baking the placeholder size into the layer.

One scope note: intermediate nodes still show the placeholder sizes they were traced with — only the input nodes report `None`. Propagating dynamic dims through layers is doable (trace twice with different placeholder sizes and diff) but it's a bigger change, so I left it out of this PR; happy to follow up if you're interested.

Includes 7 tests and a short docs section in `quick_start.md`.

### CODE REVIEW

**Comment by szmikler** [comment]
Thanks for the PR. I understand the motivation, but I am not confident about merging this as-is.

My main concern is that `None` is only reported for input nodes. Intermediate nodes still show the placeholder sizes used during tracing. I think this partial support could confuse users more than help them.

Currently the displayed shapes come directly from tracing with the input shape specified by the user. I like that the numbers shown are exactly the numbers that were traced. If users know their model accepts other input sizes, they can already pass them at runtime.

I would be more open to this if `None` was propagated through the layers, but I am not sure this can be done without adding a lot of complexity and edge cases. So for now I would prefer not to support `None` only at the input level.

**Comment by szmikler** [comment]
Hey @El3ssar, I don't think I will be working on this for now.

I appreciate the PR, but I think it should be closed, at least for now. Is this acceptable for you and your projects?
